src/Microsoft.Health.TaskManagement/JobHosting.cs

```diff
@@ -113,137 +99,120 @@ private async Task ExecuteJobAsync(JobInfo jobInfo)
-                    _logger.LogInformation("Job {JobId} completed.", jobInfo.Id);
+                    await _queueClient.CompleteJobAsync(jobInfo, false, CancellationToken.None);
                 }
                 catch (Exception completeEx)
                 {
                     _logger.LogError(completeEx, "Job {JobId} failed to complete.", jobInfo.Id);
                 }
+
+                return;
             }
-            finally
-            {
-                _activeJobsNeedKeepAlive.Remove(jobInfo.Id, out _);
-            }
-        }
 
-        private async Task KeepAliveSingleJobAsync(JobInfo jobInfo, CancellationTokenSource jobCancellationToken)
-        {
             try
             {
-                bool shouldCancel = false;
-                try
-                {
-                    bool cancelRequested = await _queueClient.KeepAliveJobAsync(jobInfo, CancellationToken.None);
-                    shouldCancel |= cancelRequested;
-                }
-                catch (JobNotExistException notExistEx)
-                {
-                    _logger.LogError(notExistEx, "Job {JobId} not exist or {RunId} not match.", jobInfo.Id, jobInfo.Version);
-                    shouldCancel = true;
-                }
-
-                if (shouldCancel && !jobCancellationToken.IsCancellationRequested)
-                {
-                    jobCancellationToken.Cancel();
-                }
+                jobInfo.Status = JobStatus.Completed;
+                await _queueClient.CompleteJobAsync(jobInfo, true, CancellationToken.None);
+                _logger.LogInformation("Job {JobId} completed.", jobInfo.Id);
             }
-            catch (Exception ex)
+            catch (Exception completeEx)
             {
-                _logger.LogError(ex, "Failed to keep alive on job {JobId}", jobInfo.Id);
+                _logger.LogError(completeEx, "Job {JobId} failed to complete.", jobInfo.Id);
             }
         }
 
-        private async Task KeepAliveJobsAsync(CancellationToken cancellationToken)
+        public static async Task<string> ExecuteJobWithHeartbeatsAsync(IQueueClient queueClient, byte queueType, long jobId, long version, Func<CancellationTokenSource, Task<string>> action, TimeSpan heartbeatPeriod, CancellationTokenSource cancellationTokenSource)
         {
-            _logger.LogInformation("Start to keep alive job message.");
+            EnsureArg.IsNotNull(queueClient, nameof(queueClient));
+            EnsureArg.IsNotNull(action, nameof(action));
 
-            while (!cancellationToken.IsCancellationRequested)
+            var jobInfo = new JobInfo { QueueType = queueType, Id = jobId, Version = version }; // ignore other data points from job info
+
+            await using (new Timer(async _ => await PutJobHeartbeatAsync(queueClient, jobInfo, cancellationTokenSource), null, TimeSpan.FromSeconds(RandomNumberGenerator.GetInt32(100) / 100.0 * heartbeatPeriod.TotalSeconds), heartbeatPeriod))
```

Comment:
Any exception in Timer would crash the process, I believe any exception in PutJobHeatbeatAsync would break the process. Repro code here:

            int count = 1;
            await using (new Timer(async _ =>
            {
                Console.WriteLine("Task.Delay(100)");
                if (count++ % 10 == 0)
                {
                    throw new Exception("Test");
                }
                await Task.Delay(100);
            }, null, 1000, 100))
            {
                Console.WriteLine("Test");
                await Task.Delay(TimeSpan.FromMinutes(20));
            }

Reply:
PutJobHeartbeatAsync has try/catch all block. This should prevent this problem.

Reply:
Here is the test code that shows how it works
[Fact]
#pragma warning disable SA1116
#pragma warning disable SA1117
#pragma warning disable SA1513
        public async Task TimerTest()
        {
            var sucess = false;
            var count = 1;
            var bad = 0;
            var good = 0;
            await using (new Timer(async _ =>
            {
                try
                {
                    if (count++ % 2 == 0)
                    {
                        throw new Exception("Test");
                    }
                    await Task.Delay(TimeSpan.FromSeconds(0.01));
                    good++;
                }
                catch
                {
                    bad++;
                }
            }, null, TimeSpan.FromSeconds(0), TimeSpan.FromSeconds(1)))
            {
                await Task.Delay(TimeSpan.FromSeconds(10));
                sucess = true;
            }
            Assert.True(sucess);
            Assert.True(good == 5, $"good: {good} == 5");
            Assert.True(bad == 5, $"bad: {bad} == 5");
        }

Reply:
I added redundant try/catch in JobHosting.PutJobHeartbeatAsync.